lib/gnat/services/service_responder.ex

```diff
@@ -0,0 +1,317 @@
+    res = case Map.get(state.subject_map, subject) do
+      nil ->
+        {nil, nil}
+      {epname, groupname} ->
+        {epname, groupname}
+    end
+    {:reply, res, state}
+  end
+
+  defp handle_ping(tail, state, rt, gnat) do
+    if should_respond(tail, state.config.name, state.instance_id) do
+      output = %WireProtocol.PingResponse{
+        name: state.config.name,
+        id: state.instance_id,
+        version: state.config.version,
+        metadata: get_in(state, [:config, :metadata])
+      }
+
+      Gnat.pub(gnat, rt, Jason.encode!(output))
+    end
+  end
+
+  defp handle_service_info(tail, state, rt, gnat) do
+    if should_respond(tail, state.config.name, state.instance_id) do
+      output = %WireProtocol.InfoResponse{
+        name: state.config.name,
+        description: state.config.description,
+        id: state.instance_id,
+        version: state.config.version,
+        endpoints: state.config.endpoints |> Enum.map(fn ep ->
+          %{
+            name: ep.name,
+            subject: ep.subject,
+            metadata: Map.get(ep, :metadata, %{}),
+            queue_group: Map.get(ep, :queue_group)
+          }
+        end)
+      }
+
+      Gnat.pub(gnat, rt, Jason.encode!(output))
+    end
+
+  end
+
+  defp handle_stats(tail, state, rt, gnat) do
+    if should_respond(tail, state.config.name, state.instance_id) do
+      output = %WireProtocol.StatsResponse{
+        name: state.config.name,
+        id: state.instance_id,
+        version: state.config.version,
+        started: state.started,
+        endpoints: Enum.map(state.config.endpoints, fn(ep) ->
+          effective_subject = derive_subscription_subject(ep)
+          {counters, last_error} = case :ets.lookup(:endpoint_stats, effective_subject) do
+            [{^effective_subject, counters, last_error}] ->
+              {counters, last_error}
+            [] ->
+              {:counters.new(3, [:atomics]), nil}
+          end
+          processing_time = :counters.get(counters, @idx_processing_time)
+          num_errors = :counters.get(counters, @idx_errors)
+          num_requests = :counters.get(counters, @idx_requests)
+          total_calls = num_errors + num_requests
+          avg = if total_calls > 0 do
+            trunc(ceil(processing_time / total_calls))
+          else
+            0
+          end
+
+          %{
+            name: ep.name,
+            subject: Map.get(ep, :subject),
+            num_requests: num_requests,
+            num_errors: num_errors,
+            last_error: last_error,
+            processing_time: processing_time,
+            average_processing_time: avg,
+            queue_group: Map.get(ep, :queue_group),
+            data: %{}
+          }
+        end)
+      }
+
+      Gnat.pub(gnat, rt, Jason.encode!(output))
+    end
+  end
+
+  @spec should_respond(list, String.t, String.t) :: boolean()
+  defp should_respond(tail, service_name, instance_id) do
```

Comment:
In your other predicate functions down below like `valid_version?` you have the question mark on the end. Do you think it would make sense to add here as well for consistency?

Reply:
Excellent point. This should've had a ?